akka-projection-core/src/test/scala/akka/projection/internal/metrics/ServiceTimeMetricSpec.scala

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.projection.internal.metrics
+
+import scala.concurrent.duration._
+
+import akka.projection.HandlerRecoveryStrategy
+import akka.projection.ProjectionId
+import akka.projection.internal.AtLeastOnce
+import akka.projection.internal.AtMostOnce
+import akka.projection.internal.ExactlyOnce
+import akka.projection.internal.FlowHandlerStrategy
+import akka.projection.internal.GroupedHandlerStrategy
+import akka.projection.internal.SingleHandlerStrategy
+import akka.projection.internal.metrics.InternalProjectionStateMetricsSpec._
+
+sealed abstract class ServiceTimeAndProcessingCountMetricSpec extends InternalProjectionStateMetricsSpec {
+  implicit var projectionId: ProjectionId = null
+
+  before {
+    projectionId = genRandomProjectionId()
+  }
+
+  def instruments(implicit projectionId: ProjectionId) = InMemInstruments.forId(projectionId)
+  val defaultNumberOfEnvelopes = 6
+
+}
+
+class ServiceTimeAndProcessingCountMetricAtLeastOnceSpec extends ServiceTimeAndProcessingCountMetricSpec {
+
+  "A metric reporting ServiceTime" must {
+    // at-least-once
+    " in `at-least-once` with singleHandler" must {
+      "reports measures for all envelopes (without afterEnvelops optimization)" in {
+        val single = Handlers.single
+        val tt = new TelemetryTester(AtLeastOnce(afterEnvelopes = Some(1)), SingleHandlerStrategy(single))
+
+        runInternal(tt.projectionState) {
+          instruments.afterProcessInvocations.get should be(defaultNumberOfEnvelopes)
+          instruments.lastServiceTimeInNanos.get() should be > (0L)
```

Comment:
Hmm, yes though it would be surprising. I mean, `System.nanoTims()` doesn't have nanosecond resolution but this particular measure (`ServiceTimeInNanos`) includes at least a string concatenation operation (from the handlers used in tests). I guess I can initialize the value to `-1L` and improve the assertion to `be >= (0L)`.

I haven't seen this assertion fail in multiple runs, though...

Reply:
yes, probably not a real problem in practise